### ISSUE TITLE
Provide official Docker image 

### ISSUE THREAD

**Issue by angelo-v** [solid/node-solid-server#1167]
To ease the deployment of NSS on an own server, there should be an official Docker image of NSS on the Docker Hub or any other public registry.

Requirements:
  * Provide a production ready Dockerfile
  * Automatic build process and publication in a docker registry (e.g. Docker Hub)
  * Tagged by version to have full control about the version to run
  * Automated tests

As luck would have it, I already prepared a functional, tested Dockerfile of NSS at [aveltens/docker-solid-server](https://github.com/angelo-v/docker-solid-server) and run it successfully [in production](https://angelo.veltens.org) for some months now. 

We could move / fork this repository to the Solid organization if you like. Further we need to set up a solid account on Docker Hub and configure the automatic build there.

**Comment by michielbdejong** [comment]
@Mitzi-Laszlo apparently https://hub.docker.com/u/solid already exists as an org, can you investigate the status of that?

**Comment by kjetilk** [comment]
@Mitzi-Laszlo, @michielbdejong, @jaxoncreed and I discussed this, and we figured that the best way to do it is probably to have a "Docker Maintainer" role within the Solid project, and since @angelo-v has already done the work, it makes sense if he occupies that role, though I may be pushing the process here. Great if you look into having such a role, and if we could have a suitable home at DockerHub for it.

**Comment by angelo-v** [comment]
Since solidproject is also taken, I just registered solidweb. If someone has a better proposal we could still change, I just wanted to secure one.

**Comment by michielbdejong** [comment]
@angelo-v so then I think all that's needed is to allow you to complete this, is to add you as committer on this github repo, right?

**Comment by angelo-v** [comment]
We could have a seperate user for the docker image  (e.g. a fork of https://github.com/angelo-v/docker-solid-server) or add the  docker stuff to a subdirectory  within this repo here.
To trigger an automated built add docker hub I need to create and connect a "service user", but I still have to figure out if this is just a regular account or something special.

**Comment by Mitzi-Laszlo** [comment]
@angelo-v thank you for the contribution!

This conversation is about the project [ASAP on Server](https://github.com/orgs/solid/projects/2) with Project Manager [Jackson Morgan](https://github.com/jaxoncreed). You can read about all the projects on https://github.com/orgs/solid/projects. 

@angelo-v you could either submit a pull request or issue for @jaxoncreed to review as Project Manager. Alternatively you could become a formal Project Contributor and work closely with the Project Manager on the whole project. The admin privileges are a reflection of the role and responsibility. What would you prefer? 

@michielbdejong I'll ask around to see who is responsible for https://hub.docker.com/u/solid as well as https://hub.docker.com/u/solidproject and get back to you if I find anything. Meanwhile I'll leave @jaxoncreed as Project Manager to decide if https://hub.docker.com/u/solidweb is the best account to use.

**Comment by kjetilk** [comment]
Actually, @Mitzi-Laszlo , I think this is broader than the project, we'd want to have Docker images of more than NSS5 at some point. I think this a role closer to the Release Manager than a specific project, though certainly, working with @jaxoncreed is how we'd do it now.

**Comment by angelo-v** [comment]
I am preferring a dedicated repository within the solid organization, e.g. `solid/docker-nss`. It would be enough for me to have the permissions for this repository.  Just placing a PR here would not be enough, because this does not allow me to setup the docker build. I could do it anyway, but then someone else has to take the "Docker maintainer" role.

**Comment by kjetilk** [comment]
Right! It could be a part of the release manager's responsibilities, but I think it makes sense to have it as a separate role in the interest of making it easier for more people to contribute.

**Comment by angelo-v** [comment]
So, how are we proceeding here?

**Comment by Mitzi-Laszlo** [comment]
@jaxoncreed as Project Manager of the ASAP on Server project would be the one to decide a route forward as well as if he would be open to @angelo-v being a formal Project Contributor for this route forward

**Comment by Mitzi-Laszlo** [comment]
@kjetilk @angelo-v What would be the project scope of the broader idea and who would be responsible for leading it?

**Comment by Mitzi-Laszlo** [comment]
If it's helpful to talk perhaps we should raise this point at the next W3C Solid Community Group call

**Comment by jaxoncreed** [comment]
@angelo-v Is it possible for you to make a pull request to Node Solid Server and we'll approve it.

**Comment by jaxoncreed** [comment]
Like I think it makes sense to have https://github.com/angelo-v/docker-solid-server integrated directly into the NSS source

**Comment by angelo-v** [comment]
Ok, who will ensure the integration with docker hub, proper tagging and regular updating?

**Comment by angelo-v** [comment]
Ideally this will be done in the same breath as a new version is released to npm

**Comment by jaxoncreed** [comment]
@kjetilk are you okay with taking this on when you push?

**Comment by gobengo** [comment]
Even better to have this (npm publish + docker push) automated by Travis or similar upon pushing a git tag.

**Comment by kjetilk** [comment]
Yeah, having Travis doing this is attractive! Anyone want to look into that?

**Comment by angelo-v** [comment]
You could just setup docker hub to automatically build git tags. You are tagging with version numbers anyway to release something, do you?

**Comment by jaxoncreed** [comment]
Docker image created here https://github.com/solid/node-solid-server/blob/065ad183224cef744cc8a7d9fa421c49ce7e7eb1/Dockerfile

**Comment by angelo-v** [comment]
> Docker image created here https://github.com/solid/node-solid-server/blob/065ad183224cef744cc8a7d9fa421c49ce7e7eb1/Dockerfile

This is not a docker image, just a Dockerfile. It does not address the issue, that there is no official docker image in any repository and no automated deployment process to publish such.

**Comment by TallTed** [comment]
@jaxoncreed @kjetilk - This should clearly be reopened. I suggest that closing should not generally be done until the issue opener (or a reasonable number of suitably knowledgeable others) concurs that a proposed solution (PR or otherwise) does in fact solve their reported issue.

**Comment by peter279k** [comment]
I think it should let this repository be a Docker image and publish on `Dockerhub`.

I think it's for @angelo-v requirement :).

**Comment by kjetilk** [comment]
Right, I agree, we should let this go all the way to have a published image.

**Comment by gobengo** [comment]
No updates in awhile on this other than moving it back to 'To do'.

It seems like the ideal next step is for someone on the solid team to use Docker Hub to set up [automated builds of docker images](https://docs.docker.com/docker-hub/builds/) ('autobuilds') from the Dockerfile in this repo.

I would do this, but I believe only someone in the 'solid' GitHub Organization can set up docker hub autobuilds for this repo.

Lastly, a detail to be worked out would be what docker hub user/organization should publish the image. Can anyone on the solid team say whether this 'solidproject' docker hub user is controlled by a member of the solid team? https://hub.docker.com/u/solidproject

**Comment by angelo-v** [comment]
> Since solidproject is also taken, I just registered solidweb. If someone has a better proposal we could still change, I just wanted to secure one.

We still do not seem to know who controlls `solidproject` :( If you want to use `solidweb` let me know

**Comment by jaxoncreed** [comment]
I would welcome a pull request to implement an official docker image. I apologize as maintaining NSS is not my full time job and I've been relatively busy over the past few weeks.

**Comment by gobengo** [comment]
@jaxoncreed I don't think any PR here can be done to host a docker image on docker hub. Someone at inrupt and/or other trusted organizations needs to click through the GUI in Docker Hub to set up autobuilds so that every git tag pushed here results in a built image tagged and hosted on docker hub.

What's really missing is the blessing of some docker hub account that can be considered 'official', e.g. it's governed like other parts of the solid project are, with some amount of transparency and review to any major decisions. It might make sense for the Self Hosting panel to shepherd that? https://github.com/solid/self-hosting-panel

@angelo-v solidweb seems fine to me as a name.

It also seems that https://hub.docker.com/u/inrupt was created a few days ago, soon after my last comment. Someone who works there may also be interested in people using an 'inrupt/node-solid-server' image.

**Comment by michielbdejong** [comment]
I requested docker as a github app for the solid org and asked @Mitzi-Laszlo  to approve it.

**Comment by kjetilk** [comment]
Ah, cool! I saw the request, and given the thread that has been here, I approved it. :-)

**Comment by michielbdejong** [comment]
https://hub.docker.com/r/nodesolidserver/node-solid-server

**Comment by michielbdejong** [comment]
@jaxoncreed @angelo-v @dmitrizagidulin @kjetilk @gobengo let me know your Docker Hub username if you have one, so I can add you to the https://hub.docker.com/r/nodesolidserver org. Automatic builds have been set up now so it will probably not require us to log in often, but currently it has me as the only org member.

**Comment by michielbdejong** [comment]
Readme updated: https://github.com/solid/node-solid-server/commit/200e0a9d57e2e9b0d38ed0b3462ad46f5f4c570c

**Comment by angelo-v** [comment]
@michielbdejong my docker hub username is `aveltens`.

Thanks for moving on with this, yet I do not consider this ticket solved. The provided docker file does not appear to be production ready to me. I put some efforts into this at https://github.com/angelo-v/docker-solid-server - regarding code as well as documentation.

If you like I prepare a PR moving stuff from there into a subfolder in this repo, but than you need to adjust the automated build again since the Dockerfile will move to a subfolder.

Regarding inrupt: Would love to see a `inrupt/pod-server` image :)

**Comment by michielbdejong** [comment]
@aveltens ah yes please do, sorry. This just uses https://github.com/solid/node-solid-server/blob/master/Dockerfile.

Subfolder sounds like a good solution!

Added you as an owner on the Docker org, once your PR is merged, you can edit the build context from `/` to whatever you call the folder, using:
https://hub.docker.com/repository/docker/nodesolidserver/node-solid-server/builds/edit

**Comment by angelo-v** [comment]
I provided the PR https://github.com/solid/node-solid-server/pull/1385
